src/hooks.rs

```diff
@@ -0,0 +1,45 @@
+//! This module provides a way to register decoding hooks for image formats not directly supported
+//! by this crate.
+
+use std::{
+    collections::HashMap,
+    io::{BufReader, Read, Seek},
+    sync::RwLock,
+};
+
+use crate::{ImageDecoder, ImageFormat, ImageResult};
+
+pub(crate) trait ReadSeek: Read + Seek {}
+impl<T: Read + Seek> ReadSeek for T {}
+
+pub(crate) static DECODING_HOOKS: RwLock<Option<HashMap<ImageFormat, DecodingHook>>> =
+    RwLock::new(None);
+
+/// A wrapper around a type-erased trait object that implements `Read` and `Seek`.
+pub struct BoxReadSeek<'a>(pub(crate) Box<dyn ReadSeek + 'a>);
+impl Read for BoxReadSeek<'_> {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        self.0.read(buf)
+    }
+}
+impl Seek for BoxReadSeek<'_> {
+    fn seek(&mut self, pos: std::io::SeekFrom) -> std::io::Result<u64> {
+        self.0.seek(pos)
+    }
+}
+
+/// A function to produce an `ImageDecoder` for a given image format.
+pub type DecodingHook = Box<
+    dyn for<'a> Fn(BufReader<BoxReadSeek<'a>>) -> ImageResult<Box<dyn ImageDecoder + 'a>>
```

Comment:
What's going on here is a bit tricky. We need a type-erased function that produces `ImageDecoder`s given a reader.  The reader itself needs to be type-erased but we also don't want to have a virtual function indirection for every `read` call. Thus we wrap the reader in a `BufReader`

Reply:
Maybe make it a concrete type that implements `BufRead` instead of exposing a `BufReader` explicitly? A custom type can always be swapped for something else. 

`BoxReadSeek` is created from `dyn BufRead`, so it's a bit unfortunate that there's double buffering going on. With some self-referential hacks it should be possible to provide a fast read implementation that reads off a slice returned by a `dyn BufRead`.

Reply:
I could add a new `struct GenericReader(BufReader<BoxReadSeek<'a>>)` type so the BufReader isn't part of the public signature. But the important thing here isn't actually the `BufRead` implementation. As you note, the inner reader will already have one. Rather, the thing we need is a `read` implementation that can be inlined everywhere the decoder calls it. It is fine if it involves a slowpath that does a virtual function call into a `Box<dyn Read>`, but decoding an image might involve millions of `read` calls, so you really don't want to be doing that every time.

A fun quirk of `BufReader` is that there _isn't_ actually double buffering going on if both have the same buffer size. That's because the inner `BufReader` [will never fill its internal buffer when fulfilling reads >= to its capacity](https://github.com/rust-lang/rust/blob/3f1be1ec7ec3d8e80beb381ee82164a0aa3ca777/library/std/src/io/buffered/bufreader.rs#L338), but the outer `BufReader` will always make exactly capacity sized reads.

Reply:
Using the buffer from the inner `dyn BufRead` would require a self referential type that could hold by the inner reader alongside a reference to its internal buffer. Might be possible with `Box::pin` and some unsafe?